roxie/udplib/udptrr.cpp

```diff
@@ -312,127 +331,87 @@ class CReceiveManager : implements IReceiveManager, public CInterface
-            currentRequester = requester;
-            requester->requestToSend(max_transfer, myNode.getIpAddress());
-            assert(timeout >= 10 && timeout <= 20000);
-            return timeout;
         }
 
-        bool noteDone(UdpSenderEntry *requester)
+        void okToSend(UdpSenderEntry *requester, unsigned slots)
         {
-            if (requester != currentRequester)
+            switch (requester->state)
             {
-                // This should not happen - I suppose it COULD if we receive a delayed message for a transfer we had earlier given up on.
-                // Best response is to ignore it if so
-                DBGLOG("Received completed message is not from current sender!");
-                // MORE - should we set currentRequester NULL here? debatable.
-                return false;
+            case flowType::request_to_send:
+                pendingRequests.remove(requester);
+                // Fall through
+            case flowType::send_completed:
+                pendingPermits.append(requester);
+                requester->state = flowType::ok_to_send;
+                break;
+            case flowType::ok_to_send:
+                // Perhaps the sender never saw our permission? Already on queue...
+                break;
+            default:
+                // Unexpected state, should never happen!
+                DBGLOG("ERROR: Unexpected state %s in enqueueRequest", flowType::name(requester->state));
+                throwUnexpected();
+                break;
             }
-            currentRequester->noteDone();
-            currentRequester = nullptr;
-            return true;
+            requester->timeStamp = msTick();
+            requester->requestToSend(slots, myNode.getIpAddress());
         }
 
-        unsigned timedOut(UdpSenderEntry *requester)
+        void noteDone(UdpSenderEntry *requester, UdpRequestToSendMsg &msg)
         {
-            // MORE - this will retry indefinitely if agent in question is dead
-            // As coded, this rescinds the permission to send that just timed out and tells the next person to have a go
-            // Thus leading to "Received completed message is not from current sender" messages the the send was in flight
-            currentRequester = nullptr;
-            if (requester->retryOnTimeout())
-                enqueueRequest(requester);
-            if (pendingRequests)
-                return sendNextOk();
-            else
-                return 5000;
+            switch (requester->state)
+            {
+            case flowType::request_to_send:
+                // A bit unexpected but will happen if our previous permission timed out and we pushed to back of the requests queue
+                pendingRequests.remove(requester);
+                break;
+            case flowType::ok_to_send:
+                pendingPermits.remove(requester);
```

Comment:
//could possibly occur incorrectly with out of order packets.  The effect will be for multiple senders to have received permits - but it will resolve.